extensions/src/main/java/org/pentaho/platform/plugin/services/metadata/SessionCachingMetadataDomainRepository.java

```diff
@@ -397,23 +422,65 @@ public void removeModel( final String domainId, final String modelId ) throws Do
   @Override
   public Set<String> getDomainIds() {
     final IPentahoSession session = PentahoSessionHolder.getSession();
-
     final String domainKey = generateDomainIdCacheKeyForSession( session );
+    return getDomainIdsHelper( session, domainKey, () -> delegate.getDomainIds() );
+  }
+
+  @Override
+  public Set<String> getMetadataDomainIds() {
+    final IPentahoSession session = PentahoSessionHolder.getSession();
+    final String domainKey = generateDomainIdCacheKeyForSession( session,
+            PentahoDatasourceType.Metadata.toString() );
+    return getDomainIdsHelper( session, domainKey, () -> delegate.getMetadataDomainIds() );
+  }
+
+  @Override
+  public Set<String> getDataSourceWizardDomainIds() {
+    final IPentahoSession session = PentahoSessionHolder.getSession();
+    final String domainKey = generateDomainIdCacheKeyForSession( session,
+            PentahoDatasourceType.DataSourceWizard.toString() );
+    return getDomainIdsHelper( session, domainKey, () -> delegate.getDataSourceWizardDomainIds() );
+  }
+
+  /**
+   * Wrapper to retrieve domains domain identifiers checking ACL. Will return results from cache, else it will
+   * query delegate. Before return, will synchronously populate the cache for the specified domain identifiers.
+   * @param session pentaho session
+   * @param domainKey key used for caching
+   * @param delegateGetDomainIds logic to retrieve domains from delegate if not in cache
+   * @return domain identifiers
+   */
+  Set<String> getDomainIdsHelper( final IPentahoSession session, String domainKey,
+                                  Supplier<Set<String>> delegateGetDomainIds ) {
+    Set<String> domainIds = getDomainIdsFromCache( domainKey, delegateGetDomainIds );
+    asyncPopulateCacheDomain( domainIds, session );
```

Comment:
[#7 professionalize code for asynchronous populate cache](https://jira.pentaho.com/browse/BACKLOG-36132)